Improve error message when Proxy object is iterated

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58302 Improve error message when Proxy object is iterated**

Differential Revision: [D28444030](https://our.internmc.facebook.com/intern/diff/D28444030)

Commit:
Improve error message when Proxy object is iterated

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28444030).

jamesr66a commented on torch/fx/proxy.py:
```diff
@@ -134,8 +134,13 @@ def iter(self, obj: 'Proxy') -> Iterator:
+                         'Consider factoring out the untraceable logic '
+                         'using torch.fx.wrap. See the torch.fx docs on'
+                         ' pytorch.org for a more detailed explanation '
+                         'of what types of control flow can be traced.')
```
I think there's another alternative: if the control flow is static (e.g. the loop trip count is based on some hyperparameter), the code can be refactored as something like:

```
for i in range(self.some_hyperparameter):
  indexed_item = proxied_value[i]
```

We might want to work something like that in as well

ansley replied:
Good idea! I moved some of that information up into the Proxy docstring to give it more visibility. Please check over the new things I added before I land the fixes

Commit:
Update on "Improve error message when Proxy object is iterated"


Differential Revision: [D28444030](https://our.internmc.facebook.com/intern/diff/D28444030)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28444030).
